[JIT/Futures] Support set_exception API on python future.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50983 [JIT/Futures] Support set_exception API on python future.**
* #50979 [torch.futures] Fix doc inconsistency about callback args
* #50978 [torch.futures] Clarify callback behavior when future is completed
* #50977 [RPC] Add option to make rref.get_type not block.

There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

rohan-varma committed:
[JIT/Futures] Add option to JIT future to take in unwrap_func from python

There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows the Future's `unwrap_func` to be set in python optionally, so users can set futures completed with an exception and the error will throw as expected. This is mostly to support the following use case in the next diff:

```
ret_fut = torch.futures.Future(unwrap_func = lambda python_result: {
    # throw exception if needed
    if isinstance(python_result, Exception):
        throw python_result
})

rpc_fut = rpc.rpc_async(...) # RPC future that times out
# Goal is to propagate RPC error to this future
rpc_fut.add_done_callback(
res => {
    # Note that ret_fut.set_result(res.wait()) won't propagate the error
    try:
        ret_fut.set_result(res.wait())
    except Exception as e:
        ret_fut.set_result(e)
}
)
```

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows the Future's `unwrap_func` to be set in python optionally, so users can set futures completed with an exception and the error will throw as expected. This is mostly to support the following use case in the next diff:

```
ret_fut = torch.futures.Future(unwrap_func = lambda python_result: {
    # throw exception if needed
    if isinstance(python_result, Exception):
        throw python_result
})

rpc_fut = rpc.rpc_async(...) # RPC future that times out
# Goal is to propagate RPC error to this future
rpc_fut.add_done_callback(
res => {
    # Note that ret_fut.set_result(res.wait()) won't propagate the error
    try:
        ret_fut.set_result(res.wait())
    except Exception as e:
        ret_fut.set_result(e)
}
)
```

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Python's Future has a `set_exception` API, does it make sense to stay consistent with that API? https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_exception

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows the Future's `unwrap_func` to be set in python optionally, so users can set futures completed with an exception and the error will throw as expected. This is mostly to support the following use case in the next diff:

```
ret_fut = torch.futures.Future(unwrap_func = lambda python_result: {
    # throw exception if needed
    if isinstance(python_result, Exception):
        throw python_result
})

rpc_fut = rpc.rpc_async(...) # RPC future that times out
# Goal is to propagate RPC error to this future
rpc_fut.add_done_callback(
res => {
    # Note that ret_fut.set_result(res.wait()) won't propagate the error
    try:
        ret_fut.set_result(res.wait())
    except Exception as e:
        ret_fut.set_result(e)
}
)
```

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma commented:
> Python's Future has a `set_exception` API, does it make sense to stay consistent with that API? https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.set_exception

Sure, this would work as well. The main goal here is to set an `unwrap_func` that would throw a python exception inline when wait()/value() is called. So set_exception can set the following python function as the unwrap_func:
```
        def processing(fut_result):
            assert isinstance(fut_result, Exception)
            raise fut_result
```

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows the Future's `unwrap_func` to be set in python optionally, so users can set futures completed with an exception and the error will throw as expected. This is mostly to support the following use case in the next diff:

```
ret_fut = torch.futures.Future(unwrap_func = lambda python_result: {
    # throw exception if needed
    if isinstance(python_result, Exception):
        throw python_result
})

rpc_fut = rpc.rpc_async(...) # RPC future that times out
# Goal is to propagate RPC error to this future
rpc_fut.add_done_callback(
res => {
    # Note that ret_fut.set_result(res.wait()) won't propagate the error
    try:
        ret_fut.set_result(res.wait())
    except Exception as e:
        ret_fut.set_result(e)
}
)
```

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows the Future's `unwrap_func` to be set in python optionally, so users can set futures completed with an exception and the error will throw as expected. This is mostly to support the following use case in the next diff:

```
ret_fut = torch.futures.Future(unwrap_func = lambda python_result: {
    # throw exception if needed
    if isinstance(python_result, Exception):
        throw python_result
})

rpc_fut = rpc.rpc_async(...) # RPC future that times out
# Goal is to propagate RPC error to this future
rpc_fut.add_done_callback(
res => {
    # Note that ret_fut.set_result(res.wait()) won't propagate the error
    try:
        ret_fut.set_result(res.wait())
    except Exception as e:
        ret_fut.set_result(e)
}
)
```

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` APi to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Add option to JIT future to take in unwrap_func from python"


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` APi to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma commented:
cc @wanchaol for JIT-side changes.

mrshenli commented on test/test_futures.py:
```diff
@@ -4,13 +4,49 @@
+        f = Future()
+        f.set_exception(value_error)  # type: ignore
+        with self.assertRaisesRegex(ValueError, "Intentional"):
+            f.value()  # type: ignore
```
what happens when calling `then()`? I assume it should still work, and the cb can retrieve the exception from the `value`/`wait` API as well?

rohan-varma replied:
It looks like the exception is only raised if we use value/wait in the then() code, otherwise the cb runs fine, which makes sense, since it doesn't use the value of the future. 

Due to https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/pybind_utils.h#L136, its RuntimeError instead of the specific error though.

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

wanchaol reviewed (approved):
JIT changes looks good to me

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]

rohan-varma commented:
Failure is unrelated to this diff:
```
Traceback (most recent call last):
  File "test_optim.py", line 386, in test_adam
    lambda weight, bias: optimizer([weight, bias], lr=1e-3)
  File "test_optim.py", line 243, in _test_basic_cases
    scheduler_constructors
  File "test_optim.py", line 126, in _test_basic_cases_template
    optimizer.step(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/optim/optimizer.py", line 89, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/optim/_multi_tensor/adam.py", line 66, in step
    loss = closure()
  File "test_optim.py", line 115, in fn
    loss.backward()
  File "/opt/conda/lib/python3.6/site-packages/torch/tensor.py", line 245, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 147, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: CUDA error: an illegal memory access was encountered
```

rohan-varma committed:
Update on "[JIT/Futures] Support set_exception API on python future."


There is currently no way to handle/propagate errors with the python-based futures API (they are raised correctly if set with an error, but this is only possible from C++).

This diff allows a `set_exception` API to allow this. This is done by setting `unwrap_func` to be a python function that raises an exception. This is mostly to support a future change that will enable `rref.rpc_async()` to be non-blocking (need this for error handling). 

Differential Revision: [D25950304](https://our.internmc.facebook.com/intern/diff/D25950304/)

[ghstack-poisoned]
